File: paint_house_recursion.cpp
//paint house recursive approach
class Solution {
public:
    int minCost(vector<vector<int>>& costs) {
        if(costs.empty() || costs.size() == 0) return 0;

        int amount = std::numeric_limits<int>::max();

        for (int j = 0; j < 3; j++){
            amount = std::min(amount, helper(costs, 0,j,0));
        }
            return amount;
    }

    int helper(vector<vector<int>>& costs, int i, int j, int amount){
        //base case
        if(i == costs.size()) return amount;


        int case1 = 0, case2 = 0;

        if(j == 0){
           case1 = helper(costs, i+1, 1, amount+costs[i][1]) ;
           case2 = helper(costs, i+1, 2, amount+costs[i][2]);
        }

        if (j == 1){
            case1 = helper(costs, i+1, 0, amount+costs[i][0]) ;
            case2 = helper(costs, i+1, 2, amount+costs[i][2]);
        }

        if(j == 2){
            case1 = helper(costs, i+1, 0, amount+costs[i][0]) ;
            case2 = helper(costs, i+1, 1, amount+costs[i][1]);
        }

        return std::min(case1, case2);

    }
};

//Time Complexity: O(3*2^n) = O(2^n)
//Space Compelxity: O(2^n)
//Time limit exceeded on leetcode

File: PaintHouseRecursion.java
// TC - O(3*2^N)
// SC - O(3*2^N)

// Constraints:
//     costs.length == n
//     costs[i].length == 3
//     1 <= n <= 100
//     1 <= costs[i][j] <= 20


public class PaintHouseRecursion {
    class Solution {
        public int minCost(int[][] costs) {
            int costFirstHouseRed = helper(costs, 0, 0, 0);
            int costFirstHouseGreen = helper(costs, 0, 1, 0);
            int costFirstHouseBlue = helper(costs, 0, 2, 0);

            return Math.min(costFirstHouseGreen, Math.min(costFirstHouseRed, costFirstHouseBlue));
        }

        private int helper(int[][] costs, int house, int color, int cost) {
            if (house >= costs.length) {
                return cost;
            }
            if (color == 0) {
                int costGreen = helper(costs, house + 1, 1, cost + costs[house][color]);
                int costBlue = helper(costs, house + 1, 2, cost + costs[house][color]);
                return Math.min(costGreen, costBlue);
            } else if (color == 1) {
                int costRed = helper(costs, house + 1, 0, cost + costs[house][color]);
                int costBlue = helper(costs, house + 1, 2, cost + costs[house][color]);
                return Math.min(costRed, costBlue);
            } else {
                int costRed = helper(costs, house + 1, 0, cost + costs[house][color]);
                int costGreen = helper(costs, house + 1, 1, cost + costs[house][color]);
                return Math.min(costRed, costGreen);
            }
        }
    }
}
